### ISSUE TITLE
Not initialising data if schema property is an object

### ISSUE THREAD

**Issue by adamreisnz** [Automattic/mongoose#6738]
**Do you want to request a *feature* or report a *bug*?**
Not sure, maybe both, maybe neither 😆 

**What is the current behavior?**
Consider the following schema (simplified):

```js
const ContactSchema = new Schema({
  name: {
    type: String,
    trim: true,
  },
  address: AddressSchema,
  customFields: {
    type: {},
    default: {},
  },
});
```

The Address child schema can look something like this (simplified):

```js
const AddressSchema = new Schema({
  city: String,
  country: String,
}, {
  _id: false,
});
```

Now if I instantiate a new contact with the following data:

```js
const contact = new Contact({
  name: 'Test',
  'address.city': 'Tauranga',
  'address.country': 'New Zealand',
  'customFields.foo': 'Foo',
  'customFields.bar': 'Bar',
});
```

The custom fields data is completely ignored, and I get the following:

```js
console.log(contact);
{
  name: 'Test',
  address: {
    city: 'Tauranga',
    country: 'New Zealand',
  }
}
```

The address nested fields are correctly parsed and interpreted as nested data, but the nested custom fields are ignored.

This is probably because customFields is not a pre-defined schema, but since the property `customFields` is defined as an object in the schema, I would have assumed that Mongoose would just accept any data there.

Thoughts?

**If the current behavior is a bug, please provide the steps to reproduce.**
See above

**What is the expected behavior?**
Expected output would be:
```js
console.log(contact);
{
  name: 'Test',
  address: {
    city: 'Tauranga',
    country: 'New Zealand',
  },
  customFields: {
    foo: 'Foo',
    bar: 'Bar',
  },
}
```

**Please mention your node.js, mongoose and MongoDB version.**
Node 8.10.x, Mongoose ^5.1.2

**Comment by lineus** [comment]
Hi @adamreisnz! It looks like this was fixed in 5.1.3 as part of changes to `document.$set` and still works correctly in 5.2.3.

### 6738.js
```js
#!/usr/bin/env node
'use strict';

const mongoose = require('mongoose');
let opts = {};

if (Number(mongoose.version.charAt(2)) > 1) {
  opts.useNewUrlParser = true;
}

mongoose.connect('mongodb://localhost:27017/test', opts);
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const addressSchema = new Schema({
  city: String,
  country: String,
}, {
  _id: false,
});

const contactSchema = new Schema({
  name: {
    type: String,
    trim: true
  },
  address: addressSchema,
  customFields: {
    type: {},
    default: {}
  }
});

const Contact = mongoose.model('contact', contactSchema);

const contact = new Contact({
  name: 'Test',
  'address.city': 'Tauranga',
  'address.country': 'New Zealand',
  'customFields.foo': 'Foo',
  'customFields.bar': 'Bar'
});

async function run() {
  console.log(`Output with ${mongoose.version}`);
  await conn.dropDatabase();
  await contact.save();
  let doc = await Contact.findOne();
  console.log(doc);
  return conn.close();
}

run();
```
### Output with 5.1.2
```
issues: ./6738.js
Output with 5.1.2
{ _id: 5b4f08f23bd84b0b3f30895f,
  name: 'Test',
  address: { city: 'Tauranga', country: 'New Zealand' },
  __v: 0 }
issues:
```
### Output with 5.1.3
```
issues: ./6738.js
Output with 5.1.3
{ _id: 5b4f08fe8fc40d0b5f7175da,
  name: 'Test',
  address: { city: 'Tauranga', country: 'New Zealand' },
  customFields: { foo: 'Foo', bar: 'Bar' },
  __v: 0 }
issues:
```
### Output with 5.2.3
```
issues: ./6738.js
Output with 5.2.3
{ _id: 5b4f090a7af6c40b7cab7f53,
  name: 'Test',
  address: { city: 'Tauranga', country: 'New Zealand' },
  customFields: { foo: 'Foo', bar: 'Bar' },
  __v: 0 }
issues:
```

## 5.1.2 workaround:

### You can change your doc creation to:
```js
const contact = new Contact({
  name: 'Test',
  'address.city': 'Tauranga',
  'address.country': 'New Zealand',
  customFields: {
    foo: 'Foo', 
    bar: 'Bar'
  }
});
``` 
### And the Output becomes:
```
issues: ./6738.js
Output with 5.1.2
{ _id: 5b4f0d74d0b4370cfd7d725c,
  name: 'Test',
  address: { city: 'Tauranga', country: 'New Zealand' },
  customFields: { foo: 'Foo', bar: 'Bar' },
  __v: 0 }
issues:
```

I think it's a reasonable idea to add a test that covers this explicit use case to ensure the behavior when using positional syntax *( ie 'customFields.bar': 'Bar )* doesn't change going forward. I believe the test(s) added in 5.1.3 only cover embedded discriminators.

**Comment by adamreisnz** [comment]
Hah, I missed the boat by one patch version it seems ;)
Thanks for checking, I will make sure to update to the latest version next time before posting.

Yes, a test would be good, there's a reason we use positional syntax here, and I did use a workaround for the time being, but I'll give it another spin now without the workaround and see how it goes.
